### PR TITLE
Add Fluid Compiler design doc

### PR DESCRIPTION
**Pull request by wangkuiyi**

### CODE REVIEW

File: doc/design/fluid_compiler.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+In the rest of this article, we talk about a special kind of
+transpiler, *Native code generator*, which takes a `ProgramDesc` and
+generates a `.cu` (or `.cc`) file, which could be built by C++
+compilers (gcc, nvcc, icc) into binaries.
+
+## Native Code Generator
+
+For the above example, the native code generator transpiler, say, the
+CUDA code generator, should generate a `main` function:
```
Comment (jacquesqiao) [inline doc/design/fluid_compiler.md]: In most case, user may need a library such as a `.so` or `.a` to use in some other program, such as a Face recognition App,  so do we also need to generate these libraries at the same time?

File: doc/design/fluid_compiler.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+and the definitions of functions `fluid_cuda_read`,
+`fluid_cuda_create_tensor`, and `fluid_cuda_mult`.  Please be aware
+that each function could just define a C++ instance of an operator and
+run it.  For example
+
+```c++
+paddle::Tensor fluid_cuda_read(...) {
+  paddle::Tensor t;
+  paddle::operator::Read r(&t, ...);
```
Comment (reyoung) [inline doc/design/fluid_compiler.md]: Should we copy the `operator::Read` files (`read_op.cc`/`read_op.cu`/`read_op.h`) to the generated project. Or just `read_op.a/.so`?

File: doc/design/fluid_compiler.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
+# PaddlePaddle Fluid: Towards a Compiled Programming Language
+
+As described in [fluid.md](fluid.md), when a Fluid application program
+runs, it generates a `ProgramDesc` protobuf message as an intermediate
+representation of itself.  The C++ class `Executor` can run this
+protobuf message as an interpreter.  This article describes the Fluid
+compiler.
+
+![](fluid-compiler.png)
```
Comment (helinwang) [inline doc/design/fluid_compiler.md]: Do we need `Executor` in this graph, `Executor` seems unrelated to this PR.

File: doc/design/fluid_compiler.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  Y = fluid.mult(X, W)
+}
+```
+
+This program consists of a [block](block.md) of three operators --
+`read`, `assign`, and `mult`.  Its `ProgramDesc` message looks like
+the following
+
+```protobuf
+message ProgramDesc {
```
Comment (helinwang) [inline doc/design/fluid_compiler.md]: From my understanding the ProgramDesc is a intermediate representation (IR), currently we have Python as frontend that generates the IR, and this PR discusses a cpp code backend.

I think having Python as a frontend is a huge pain. In my opinion the benefit of Python in the machine learning field is:

1. libraries such as numpy
1. Python native program control primitives such as `for`, `if` to control the training process, the researchers are familiar with them.

In our case we are benefit from neither of these two points:

1. the transpiled code can not use numpy.
1. the transpiled code can not use Python native program control primitives.

And we are trapped in the Python grammar.

I think a better way is to invent our own language.

File: doc/design/fluid_compiler.md (hunk 1 above)
Comment (wangkuiyi) [reply doc/design/fluid_compiler.md]: Yes. We can. It looks to me that `paddle/framework.a` is the runtime of the Fluid compiler, just like libgcc.a for GCC.

File: doc/design/fluid_compiler.md (hunk 2 above)
Comment (wangkuiyi) [reply doc/design/fluid_compiler.md]: I think we need to copy the source code into the output.{cc,cu} file, which will then be built by NVCC or ICC or other compilers.

File: doc/design/fluid_compiler.md (hunk 4 above)
Comment (wangkuiyi) [reply doc/design/fluid_compiler.md]: I agree and I believe that a new language is the future.

File: doc/design/fluid_compiler.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+and the definitions of functions `fluid_cuda_read`,
+`fluid_cuda_create_tensor`, and `fluid_cuda_mult`.  Please be aware
+that each function could just define a C++ instance of an operator and
+run it.  For example
+
+```c++
+paddle::Tensor fluid_cuda_read(...) {
+  paddle::Tensor t;
+  paddle::operator::Read r(&t, ...);
+  r.Run();
```
Comment (sidgoyal78) [inline doc/design/fluid_compiler.md]: This might be a very basic question, but from this example, I see `r.Run()`. Is this bringing execution into picture along with the compilation? (Maybe we instead have a Run() method, which is called by the executor later?, or maybe i misunderstood).

Comment (sidgoyal78) [reply doc/design/fluid_compiler.md]: Oh nevermind, I mixed up the 2 things. I think post this code generation, the executor will indeed run these as usual i guess.

File: doc/design/fluid.md [hunk 6]
```diff
@@ -105,18 +105,10 @@ There are two ways to execute a Fluid program.  When a program is executed, it c
 
 There is a C++ class [`Executor`](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/executor.h), which runs a `ProgramDesc`, similar to how an interpreter runs a Python program.
 
-Fluid is moving towards the direction of a compiler, which is explain in more detail later in this article.
+Fluid is moving towards the direction of a compiler, which is explain in [fluid_compiler.md](fluid_compiler.md).
```
Comment (abhinavarora) [inline doc/design/fluid.md]: explain -> **explained**

File: doc/design/fluid_compiler.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  return t;
+}
+```
+
+For computational operators that have multiple *kernels*, each for a
```
Comment (abhinavarora) [inline doc/design/fluid_compiler.md]: Should we also consider the possibility of having a default fallback device in case some operator cannot run on a device. For example, we might have some CPU only operators in  that case our transpiler should make sure that it generates CPU code for that op even though the rest of the native code might be CUDA code?

Comment (dzhwinter) [reply doc/design/fluid_compiler.md]: @abhinavarora 
Yes. We provide the kernel selection and kernel fallback mechanism. 
https://github.com/PaddlePaddle/Paddle/blob/eee62648cf075e6520405d6ae18b039a357d40e8/paddle/framework/data_transform.h#L33-L40
If the target machine does not have the target device, it will try to use naive implement kernel (say CPU kernel) instead of terminated.
In the fluid overview design, there will be a runtime `.a` link to target program. In my view, the runtime library will solve the fallback problem, not the transpiler does.
